logr.go

```diff
@@ -279,7 +286,12 @@ func (l Logger) Error(err error, msg string, keysAndValues ...interface{}) {
 	if withHelper, ok := l.sink.(CallStackHelperLogSink); ok {
 		withHelper.GetCallStackHelper()()
 	}
-	l.sink.Error(err, msg, keysAndValues...)
+	sink := l.sink
+	contextKeysAndValues := l.valuesFromContext()
```

Comment:
It seems a shame to do all this (allocations) on every log call, when we know that the returned slice only changes when `WithContext()` or `WithContextValues()` is called.

Perhaps worth investigating a followup optimization which does this only when needed.  E.g.

```go
// ContextLogSink is an optional interface which LogSinks can implement to optimize
// the generation of key-value pairs from a context.Context.  LogSinks which implement
// this are expected to store a set of key-value pairs, just like WithValues but distinct, 
// which are logged with every log line.
type ContextLogSink interface {
    // This is similar to LogSink.WithValues except that the LogSink must discard
    // any saved context-value and replace them with this new set.  This will be called 
    // whenever the Logger's WithContext or WithContextValues methods are called.
    WithContextValues(keysAndValues ...interface{}) LogSink
}
```

Then Info/Error do something like:

```go
	sink := l.sink
	if _, ok :- sink.(ContextLogSink); !ok {
		// The sink does not store context values itself, so we have to
		// generate them.
		contextKeysAndValues := l.valuesFromContext()
		if len(contextKeysAndValues) > 0 {
			sink = sink.WithValues(contextKeysAndValues...)
		}
	}
	sink.Info(l.level, msg, keysAndValues...)
```

Then WithContext:

```go
	l.ctx = ctx
	if cls, ok := l.sink.(ContextLogSink); !ok {
		l.sink = cls.WithContextValues(l.valuesFromContext())
	}
	return l
```

We could go further and store the results of `valuesFromContext()` in Logger so we don't have to regenerate it on every Info/Error (but still have to call sink.WithValues()), but that would add more to Logger, so maybe not worth it (or deserves its own discussion).

Reply:
Both is worth investigating, although I agree that making Logger larger is probably not going to be worth it (affects everything that copies it, does not safe that much when there aren't many context keys, not needed when LogSinks implement `WithContextValues`.

When you [switched to `Logger` being a struct](https://github.com/go-logr/logr/commit/bb231ff675f32d58bb1eb2ebe42756339b560feb), you wrote in the commit message that "this [= variadic slice setup] is a candidate for Go to optimize". Was that based on the expectation that Go would already do that optimization, or that a future Go release would add support for such an optimization?

I benchmarked recently whether the `if loggerV := logger.V(5); loggerV.Enabled` is justified when performance matters and found that it is because that optimization doesn't seem to happen, i.e. `logger.V(5).Info` is more expensive.

We cannot go back to `Logger` being an interface, I'm just trying to understand what it brings us and what might still be needed to make it better :sweat_smile:

Reply:
/hold

I need to try this out in k/k first, and it looks like the basic premise of the current proposal ("it is possible to extract values from a context") isn't true for OpenTelemetry because it doesn't export its keys. It only has accessor methods. Over in https://github.com/kubernetes/klog/issues/356 I've outlined a potential solution, but it needs further discussion about the desired outcome.